[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111111
* #111120
* #111110
* #111109
* #111275
* #111107
* #111106

As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

fegin committed:
[state_dict][6/N] Add a fine tuning e2e test case for distributed.state_dict and DCP

As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][6/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][6/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/checkpoint/e2e/test_fine_tuning.py:
```diff
@@ -0,0 +1,192 @@
+class TestFineTuning(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(4, torch.cuda.device_count())
```
why wouldn't it work on something like 8 GPUs?

rohan-varma replied:
In this code, if we have 8 GPUs, it will use only 4 though, right?

fegin replied:
yup, I can change it to `max(2, torch.cuda.device_count())` so that we can utilize 8 GPUs.

rohan-varma commented on test/distributed/checkpoint/e2e/test_fine_tuning.py:
```diff
@@ -0,0 +1,192 @@
+            pretrain_state_dict, _ = get_state_dict(
+                model,
+                submodules={model.pretrain},
+                options=StateDictOptions(keep_submodule_prefixes=False)
```
what does `keep_submodule_prefixes` mean? How would a user decide / know whether to use this flag or not?

fegin replied:
This is a good question. It mean whether the keys of  the returned state_dict should keep the submodule prefixes. For this example, if we set the flag to True (the default one), the returned state_dict will be `{"pretrain.layer1.weight" ...}`. And if we set the flag to False, the returned state_dict will be `{"layer1.weight", ...}`.

How would a user knows? Unfortunately, this depends on what is saved in the checkpoint. If the checkpoint already has the prefix, the user should set it to True otherwise False.

rohan-varma commented on test/distributed/checkpoint/e2e/test_fine_tuning.py:
```diff
@@ -0,0 +1,192 @@
+            # Load pretrain submodules checkpoint
+            pretrain_state_dict, _ = get_state_dict(
+                model,
+                submodules={model.pretrain},
```
In general, accessing the pretrained part might not be this easy, i.e. in cases like LoRA. We might have to support some more expressivity here. Still need to think about it and understand the PEFT use cases.

rohan-varma commented on test/distributed/checkpoint/e2e/test_fine_tuning.py:
```diff
@@ -0,0 +1,192 @@
+    def finetune(self, pretrain_dir: str, finetune_dir: str) -> None:
+        model = FineTuningModel().cuda()
+        # TODO: make the parallelism more complicated, e.g., using 2D + DDP.
+        model = FSDP(model, use_orig_params=True)
```
should we test with shardedStateDict?

rohan-varma commented on test/distributed/checkpoint/e2e/test_fine_tuning.py:
```diff
@@ -0,0 +1,192 @@
+
+            try:
+                # Load training submodules checkpoint
+                model_state_dict, optim_state_dict = get_state_dict(
```
This might just be me, but this sort of pattern of having to get the state_dict and then pass it into DCP is a bit confusing, because it almost reads as if I'm saving a state_dict here, but I'm actually just creating a holder and DCP will inplace load into it. 

I know that DCP has to do this, but could we maybe support a higher-level `load_state_dict` API that manages this for users?

fegin replied:
That's why I changed the name to `get_state_dict`, which means getting the current state_dict. It can be used for saving but can also be used for loading. The full story for loading is getting the current state dict to represent the current sharding of the model (hence `get_state_dict`) and DCP load the tensor data based on the current sharding (hence `load_state_dict`).

It is not easy to merge this get_state_dict to DCP load_state_dict. For example, we may want to have a centralize checkpoint for all different fine tuning model. So each model will require to read 2 different checkpoints, like this example. Having only one `load_state_dict` will make things much harder to achieve.

rohan-varma commented on test/distributed/checkpoint/e2e/test_fine_tuning.py:
```diff
@@ -0,0 +1,192 @@
+                    optim_state_dict=optim_state_dict,
+                    options=StateDictOptions(strict=False),
+                )
+            except KeyError:
```
I don't quite understand the explanation below - why do we hit a KeyError here on non-zero iteration? Shouldn't we be able to load the finetuning part in the middle of training?

rohan-varma replied:
Oh nvm, this is because we haven't done any finetuning yet, so there isn't a finetune state_dict to load. Unsure if there's a way to make this more clear.

fegin replied:
Correct, DCP does not have an API to check, at this moment. So we have to rely on try/except.

rohan-varma commented on test/distributed/checkpoint/e2e/test_fine_tuning.py:
```diff
@@ -0,0 +1,192 @@
+            set_state_dict(
+                model,
+                model_state_dict={model.pretrain: pretrain_state_dict},
+                options=StateDictOptions(strict=False),
```
this may be follow up work, but since strict=False is needed so we don't error out on the finetuning parts, how do we validate correctness that the pretrain part was loaded correctly? is it worth adding features to this API to check correctness on submodules?

fegin replied:
Yup, this is a good question. I noticed the errors can be silently ignored due to `strict=False`. Need to think about a way to verify.

fegin replied:
We can let `set_state_dict` to check if all the parameters in the submodules are NOT in the missing list. However, this is not always accurate if we have ignore parameters.

rohan-varma reviewed (commented):
I think this makes sense, but there's one thing I can't quite figure out for the PEFT use case. we can discuss more offline.

When I load in a large model, it seems here that I need to initialize the LM, adapt to whatever finetuning method, then wrap with FSDP, then load the LM checkpoint, then load the finetune checkpoint. 

When I switch finetuning techniques (i.e. in inference), it seems like I'd need to re-do all the steps (i.e. would need to re-wrap with FSDP). It doesn't seem like we're able to take an FSDP wrapped LM with a checkpoint already loaded, adapt it to a new PEFT technique, and load in only that delta. We have the large LM checkpoint load everytime?

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

wz337 reviewed (approved):
Accepted in phabricator.

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin committed:
Update on "[state_dict][7/N] Add a fine tuning e2e test case for distributed.state_dict and DCP"


As title

Differential Revision: [D50209732](https://our.internmc.facebook.com/intern/diff/D50209732/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
